### PR TITLE
avrlibc: build faster, make attiny devices work

### PR DESCRIPTION
**Pull request by pennae**
###### Description of changes

unlike https://github.com/NixOS/nixpkgs/issues/48205#issuecomment-863881713 suggests there's still a few problems left. programs for mega devices currently build fine out of the box, but for tiny devices the compiler isn't able to find crt objects and platform support libraries (which include such unimportant details as the interrupt vector table). this adds a setup hook to avrlibc that does the work the previously mentioned bug report commenters suggested to handle with a wrapper.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/misc/avr/libc/setup-hook.sh [hunk 1]
```diff
@@ -0,0 +1 @@
+export NIX_CFLAGS_COMPILE_avr=-B@out@/avr/lib
```
Comment (bjornfor) [inline pkgs/development/misc/avr/libc/setup-hook.sh:1]: Is it safe to *set* the value (as opposed to append/prepend)?

File: pkgs/development/misc/avr/libc/default.nix [hunk 2]
```diff
@@ -17,6 +17,10 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
 
+  # set avr-specific CFLAGS. this is necessary to properly compile code for attiny devices,
```
Comment (bjornfor) [inline pkgs/development/misc/avr/libc/default.nix:20]: Nit: please capitalize first word in sentences. (Incl. commit message.)

**Comment by emilytrau** [comment]
could this be related to why attiny is broken? https://github.com/NixOS/nixpkgs/blob/0c4800d579af4ed98ecc47d464a5e7b0870c4b1f/pkgs/build-support/bintools-wrapper/default.nix#L299

**Comment by emilytrau** [comment]
sorry this slipped through the cracks. @pennae is this ready for merge?

**Comment by WizardUli** [comment]
> it worked when we last tested it, but we've since given up entirely on avr devices (and _all_ of our programmers broke, so we can't even test the outputs any more) :/

AFAIK AVR architecture is so legacy (and with exuberant prices) that even GCC is considering dropping support altogether in newer versions.

**Comment by emilytrau** [comment]
I'm mainly looking into avr again since #270986 made me realise it's gotten a little (more) rusty. This looks like a good change I'd like to get in, I'm willing to rebase and test once the improvements in #271787 have gone through :)

**Review by emilytrau — APPROVED** [review]

**Comment by emilytrau** [comment]
Turns out I didn't test well enough sorry :(. I made a test to check all QMK platforms in #273047 and it revealed this change was not compatible with all AVR platforms and needed to revert b2844f89d10d4a2665aed231144950c354f275a6 .

**Pull request merged**